autonat.go

```diff
@@ -135,50 +135,78 @@ func (as *AmbientAutoNAT) autodetect() {
 	}
 
 	cli := NewAutoNATClient(as.host, as.getAddrs)
-	failures := 0
-
-	for _, pi := range peers {
-		ctx, cancel := context.WithTimeout(as.ctx, AutoNATRequestTimeout)
-		as.host.Peerstore().AddAddrs(pi.ID, pi.Addrs, pstore.TempAddrTTL)
-		a, err := cli.DialBack(ctx, pi.ID)
-		cancel()
-
-		switch {
-		case err == nil:
-			log.Debugf("NAT status is public; address through %s: %s", pi.ID.Pretty(), a.String())
-			as.mx.Lock()
-			as.addr = a
-			as.status = NATStatusPublic
-			as.confidence = 0
-			as.mx.Unlock()
-			return
+	ctx, cancel := context.WithTimeout(as.ctx, AutoNATRequestTimeout)
+	defer cancel()
+
+	var mx sync.Mutex
+	var pubaddr ma.Multiaddr
+	private := 0
+	public := 0
+
+	probe := 3 - as.confidence
+	if probe == 0 {
+		probe = 1
+	}
+	if probe > len(peers) {
+		probe = len(peers)
+	}
 
-		case IsDialError(err):
-			log.Debugf("dial error through %s: %s", pi.ID.Pretty(), err.Error())
-			failures++
-			if failures >= 3 || as.confidence >= 3 { // 3 times is enemy action
-				log.Debugf("NAT status is private")
-				as.mx.Lock()
-				as.status = NATStatusPrivate
-				as.confidence = 3
-				as.mx.Unlock()
-				return
+	var wg sync.WaitGroup
+
+	for _, pi := range peers[:probe] {
+		wg.Add(1)
+		go func(pi pstore.PeerInfo) {
+			as.host.Peerstore().AddAddrs(pi.ID, pi.Addrs, pstore.TempAddrTTL)
+			a, err := cli.DialBack(ctx, pi.ID)
+
+			switch {
+			case err == nil:
+				log.Debugf("Dialback through %s successful; public address is %s", pi.ID.Pretty(), a.String())
+				mx.Lock()
+				public++
+				pubaddr = a
+				mx.Unlock()
+
+			case IsDialError(err):
+				log.Debugf("Dialback through %s failed", pi.ID.Pretty())
+				mx.Lock()
+				private++
+				mx.Unlock()
+
+			default:
+				log.Debugf("Dialback error through %s: %s", pi.ID.Pretty(), err)
 			}
 
-		default:
-			log.Debugf("Error dialing through %s: %s", pi.ID.Pretty(), err.Error())
-		}
+			wg.Done()
+		}(pi)
 	}
 
+	wg.Wait()
+
 	as.mx.Lock()
-	if failures > 0 {
-		as.status = NATStatusPrivate
-		as.confidence++
+	if public > 0 {
+		log.Debugf("NAT status is public")
+		if as.status == NATStatusPrivate {
+			as.confidence = 0
+		} else if as.confidence < 3 {
+			as.confidence++
+		}
+		as.status = NATStatusPublic
+		as.addr = pubaddr
+	} else if private > 0 {
 		log.Debugf("NAT status is private")
+		if as.status == NATStatusPublic {
+			as.confidence = 0
+		} else if as.confidence < 3 {
+			as.confidence++
+		}
+		as.status = NATStatusPrivate
+		as.addr = nil
 	} else {
+		log.Debugf("NAT status is unknown")
```

Comment:
Is it ok for a single round with unresponsive peers to reset our NAT status entirely? I think we need some error tolerance here.

Reply:
I'd decrement the confidence and retain our previous status, until confidence drops to zero and then we can swith to Unknown?

Reply:
ok, that's reasonable.

Reply:
Done - won't flip to unknown if we have any confidence, but rather reduce the confidence and wait for the next round.
It will only become unknown again if the confidence drops to 0.